File: src/main/java/rars/venus/registers/ControlAndStatusWindow.java
package rars.venus.registers;

import org.jetbrains.annotations.NotNull;
import rars.Globals;
import rars.settings.BoolSetting;
import rars.venus.NumberDisplayBaseChooser;
import rars.venus.VenusUI;

import static rars.Globals.BOOL_SETTINGS;

public final class ControlAndStatusWindow extends RegisterBlockWindowBase {
    /**
     * The tips to show when hovering over the names of the registers
     */
    // TODO: Maintain order if any new CSRs are added
    private static final String[] regToolTips = {
        /* ustatus" */ "Interrupt status information (set the lowest bit to enable exceptions)",
        /* fflags */ "The accumulated floating point flags",
        /* frm */ "Rounding mode for floating point operatations",
        /* fcsr */ "Both frm and fflags",
        /* uie */ "Finer control for which interrupts are enabled",
        /* utvec */ "The base address of the interrupt handler",
        /* uscratch */ "Scratch for processing inside the interrupt handler",
        /* uepc */ "PC at the time the interrupt was triggered",
        /* ucause */ "Cause of the interrupt (top bit is interrupt vs trap)",
        /* utval */ "Value associated with the cause",
        /* uip */ "Shows if any interrupt is pending and what type",
        /* cycle */ "Number of clock cycles executed",
        /* time */ "Time since some time in the past (Milliseconds since 1/1/1970 in RARS)",
        /* instret */ "Instructions retired (same as cycle in RARS)",
        /* cycleh */ "High 32 bits of cycle",
        /* timeh */ "High 32 bits of time",
        /* instreth */ "High 32 bits of instret"
    };

    public ControlAndStatusWindow(final @NotNull VenusUI mainUI) {
        super(Globals.CS_REGISTER_FILE, regToolTips, "Current 32 bit value", mainUI);
    }

    @Override
    protected @NotNull String formatRegisterValue(final long value, final int base) {
        if (BOOL_SETTINGS.getSetting(BoolSetting.RV64_ENABLED)) {
            return NumberDisplayBaseChooser.formatNumber(value, base);
        } else {
            return NumberDisplayBaseChooser.formatNumber((int) value, base);
        }
    }
}

File: src/main/java/rars/venus/registers/ControlAndStatusWindow.kt
package rars.venus.registers

import rars.riscv.hardware.registerFiles.CSRegisterFile
import rars.settings.AllSettings
import rars.settings.BoolSetting
import rars.venus.NumberDisplayBaseChooser
import rars.venus.VenusUI

class ControlAndStatusWindow(
    csRegisterFile: CSRegisterFile,
    mainUI: VenusUI,
    settings: AllSettings,
) : RegisterBlockWindowBase(
    csRegisterFile,
    regToolTips,
    "Current 32 bit value",
    mainUI,
    settings
) {
    override fun formatRegisterValue(value: Long, base: Int): String =
        if (settings.boolSettings.getSetting(BoolSetting.RV64_ENABLED)) {
            NumberDisplayBaseChooser.formatNumber(value, base)
        } else {
            NumberDisplayBaseChooser.formatNumber(value.toInt(), base)
        }

    companion object {
        // Maintain order if any new CSRs are added
        /** The tips to show when hovering over the names of the registers */
        private val regToolTips = arrayOf(
            "Interrupt status information (set the lowest bit to enable exceptions)", // ustatus
            "The accumulated floating point flags", // fflags
            "Rounding mode for floating point operatations", // frm
            "Both frm and fflags", // fcsr
            "Finer control for which interrupts are enabled", // uie
            "The base address of the interrupt handler", // utvec
            "Scratch for processing inside the interrupt handler", // uscratch
            "PC at the time the interrupt was triggered", // uepc
            "Cause of the interrupt (top bit is interrupt vs trap)", // ucause
            "Value associated with the cause", // utval
            "Shows if any interrupt is pending and what type", // uip
            "Number of clock cycles executed", // cycle
            "Time since some time in the past (Milliseconds since 1/1/1970 in RARS)", // time
            "Instructions retired (same as cycle in RARS)", // instret
            "High 32 bits of cycle", // cycleh
            "High 32 bits of time", // timeh
            "High 32 bits of instret" // instreth
        )
    }
}
